### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by meluru**
API documentation is still WIP but it would be great if you could already check the installation, setup and example usage steps if they make sense.

### CODE REVIEW

**Comment by meluru** [comment]
Api documentation updated, let me know if I should add something more to it.

File: docs/source/index.rst [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+Current plugin
+--------------
+
+Very useful feature of Flask-PluginEngine is a ``current_plugin`` global value. It's a proxy to currently active plugin.
+
+Example plugin functionality
+----------------------------
+
+Now let's give our plugin some functionality to show how you can use the plugin. In this example we will render plugin's template from our application.
```
Comment (nop33) [inline docs/source/index.rst]: "Now let's give our plugin some functionality to show how you can use ~the plugin~ **it.** In this example we will render **the** plugin's template from our application."

File: docs/source/index.rst [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+Let's create the templates folder with a simple Jinja template. For example displaying plugin's title and description::
```
Comment (nop33) [inline docs/source/index.rst]: "Let's create the templates folder with a simple Jinja template ~. For example displaying~ **to display for example the** plugin's title and description"

File: docs/source/index.rst [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+    {% extends 'base.html' %}
+
+    {% block content %}
+        <h1>{{ plugin.title }}</h1>
+        <div>{{ plugin.description }}</div>
+    {% endblock %}
+
+If you documented plugin's class now you can access it as a ``name`` and ``description`` class properties.
```
Comment (nop33) [inline docs/source/index.rst]: I am afraid I don't understand what this means.

File: docs/source/index.rst [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+Notice that the template extends a base.html. We are going to use application's base template.
```
Comment (nop33) [inline docs/source/index.rst]: "...going to use **the** application's..."

File: docs/source/index.rst [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+
+Let's create a blueprint in our plugin::
```
Comment (nop33) [inline docs/source/index.rst]: which file?

File: docs/source/index.rst [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+
+    plugin_blueprint = PluginBlueprint('example', __name__)
+
+In order to create a blueprint of a plugin we need to create PluginBlueprint instance.
```
Comment (nop33) [inline docs/source/index.rst]: "...we need to create **a** PluginBlueprint instance."

File: docs/source/index.rst [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+Then we can create a function rendering the template::
+
+    @plugin_blueprint.route('/plugin')
+    def hello_plugin():
+    return render_template('example_plugin:index.html', plugin=current_plugin)
+
+We can render any plugin template inside our application specifying name of the plugin before template name in ``render_template`` function, like above.
```
Comment (nop33) [inline docs/source/index.rst]: "...specifying **the** name of the plugin before **the** template name..."

File: docs/source/index.rst [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+We are also passing a current_plugin object to the template.
+
+
+Now in the plugin class we can create a method returning plugin's blueprint::
```
Comment (nop33) [inline docs/source/index.rst]: "Now **, (note the comma)** in the plugin class we can create a method returning **the** plugin’s blueprint"

File: docs/source/index.rst [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+        """ExamplePlugin
+
+        Just an example plugin
+        """
+
+        def get_blueprint(self):
+            return plugin_blueprint
+
+
+To get more about information see `Plugin`_ API reference.
```
Comment (nop33) [inline docs/source/index.rst]: "To get more about information see **the** `Plugin`_ API reference."

File: docs/source/index.rst [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+
+Minimal Application set up
```
Comment (nop33) [inline docs/source/index.rst]: set up or setup? I think the second is more common, no?

File: docs/source/index.rst [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+--------------------------
+
+To create Flask application using Flask-PluginEngine we need to define it as PluginFlask application::
```
Comment (nop33) [inline docs/source/index.rst]: "To create **a** Flask application..."

File: docs/source/index.rst [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+    app = PluginFlask(__name__)
+
+    @app.route("/")
+    def hello():
+        return "Hello world!"
+
+    if __name__ == "__main__":
+        app.run()
+
+Now we need to set up the configuration: namespace of plugins we will be using with the application and plugins we want to use.
```
Comment (nop33) [inline docs/source/index.rst]: I am a bit lost with the syntax but (if I understood it correctly) I think this is cleaner:
"Now we need to set up the configuration, **specifically the** namespace of **the** plugins ~we will be using with the application~ and **the** plugins we want to use."

File: docs/source/index.rst [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+
+    app.config['PLUGINENGINE_NAMESPACE'] = 'example'
+    app.config['PLUGINENGINE_PLUGINS'] = ['example_plugin']
+
+Next we need to create an instance of PluginEngine for our app and load the plugins::
+
+    plugin_engine = PluginEngine(app=app)
+    plugin_engine.load_plugins(app)
+
+Then we can access loaded plugins by calling :func:`get_active_plugins` method, it will return a dictionary.
```
Comment (nop33) [inline docs/source/index.rst]: "Then **, (comma)** we can access **the** loaded plugins by calling **the** :func:`get_active_plugins` method **which**  ~it~ will return a dictionary" (maybe add "...will return a dictionary _containing ....._"?)

File: docs/source/index.rst [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+If we want to check if all of the plugins were loaded correclty we can also call :func:`get_failed_plugins` method that will return a dictionary with plugins that failed to load::
```
Comment (nop33) [inline docs/source/index.rst]: "~If we want~ **T**o check if all of the plugins were loaded correc**tl**y (typo)"

Comment (nop33) [inline docs/source/index.rst]: "... we can also call **the** :func..."

File: docs/source/index.rst [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+
+    active_plugins = plugin_engine.get_active_plugins(app=app).values()
+
+
+Example application functionality
+---------------------------------
+
+To understand the possibilities of Flask-PluginEngine we will create a Jinja template, where we will list all the plugins we are using.
```
Comment (nop33) [inline docs/source/index.rst]: create a Jinja template **(remove comma)** where we will

File: docs/source/index.rst [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+Let's create two templates, the base one (base.html), for example::
```
Comment (nop33) [inline docs/source/index.rst]: I'd reword it a bit to:
"Let's create two templates, `base.html` and `index.html`: " and add comments above each code block below, like this:
```jinja
{# base.html #}

<!DOCTYPE html>
 
<html lang="en">
    <head>
        ...
```
```jinja
{# index.html #}

<!DOCTYPE html>
 
<html lang="en">
    <head>
        ...
```

File: docs/source/index.rst [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+        {% endfor %}
+    {% endblock %}
+
+And let's pass the plugins, when rendering the template in our application::
+
+    @app.route("/")
+    def hello():
+    return render_template('index.html', plugins=active_plugins)
+
+Now what we also should do is register the blueprints of all our plugins, for instance::
```
Comment (nop33) [inline docs/source/index.rst]: "Now what we **should also** do is **to** register the blueprints of all our plugins..."

File: docs/source/index.rst [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+
+    @app.route("/")
+    def hello():
+        return render_template('index.html', plugins=active_plugins)
+
+
+    if __name__ == "__main__":
+        app.run()
+
+Now when we will go to the index page of our application we will be able to access plugin's template.
```
Comment (nop33) [inline docs/source/index.rst]: "Now, when we ~will~ go to the index page of our application we ~will be able to~ **can** access **the** plugin's template."

File: docs/source/index.rst [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+
+You can find the application in the source code in example folder.
```
Comment (nop33) [inline docs/source/index.rst]: "You can find the source code of the application in the example folder"

File: docs/source/index.rst [hunk 20]
```diff
@@ -0,0 +1,275 @@
...
+    :members: init
+
+    .. automethod:: plugin_context()
+    .. classmethod:: instance()
+
+        The Plugin instance used by the current app
+
+    .. classmethod:: title()
+
+        Plugin's title from the docstring
```
Comment (nop33) [inline docs/source/index.rst]: Isn't this supposed to be fetched automatically?

File: docs/source/index.rst (hunk 10 above)
Comment (meluru) [inline docs/source/index.rst]: Actually setup is a noun, while set up is a verb. So yes, it should be setup

File: docs/source/index.rst (hunk 20 above)
Comment (meluru) [inline docs/source/index.rst]: It should be, but due to the decorator problem I had to write it

**Comment by meluru** [comment]
Updated

File: docs/source/index.rst [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+
+    @app.route("/")
+    def hello():
+        return render_template('index.html', plugins=active_plugins)
+
+
+    if __name__ == "__main__":
+        app.run()
+
+Now when we go to the index page of our application we can access the plugin's template.
```
Comment (mvidalgarcia) [inline docs/source/index.rst]: Maybe it's worth to mention here that you can also go to `/plugin` and access the template of the plugin itself via its blueprint so that you prove that the stuff implemented in the plugin can be used.

File: docs/source/index.rst [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+
+You can find the source code of the application in the example folder.
```
Comment (mvidalgarcia) [inline docs/source/index.rst]: I'd appreciate here a link to the example folder.

File: docs/source/index.rst [hunk 23]
```diff
@@ -0,0 +1,291 @@
...
+    plugin_engine = PluginEngine(app=app)
+    plugin_engine.load_plugins(app)
+
+Then, we can access the loaded plugins by calling the :func:`get_active_plugins` method, which will return a dictionary containing the active plugins.
+To check if all of the plugins were loaded correctly we can also call the :func:`get_failed_plugins` method that will return a dictionary with plugins that failed to load::
+
+    active_plugins = plugin_engine.get_active_plugins(app=app).values()
+
+
+Example application functionality
```
Comment (mvidalgarcia) [inline docs/source/index.rst]: When I started reading this section I didn't know where to write the fragments of code explained and I struggled a bit until I check the example folder. In my opinion would be useful here to either explain that the plugin will be in a separated folder and the app itself in a different one or show again a directory tree pointing where the code in this section should be written in.

**Comment by meluru** [comment]
Updated

File: docs/source/index.rst [hunk 24]
```diff
@@ -26,17 +26,29 @@ You can install Flask-PluginEngine with the following command::
 Creating a plugin
 -----------------
 
-Creating a plugin is simple. First of all you need to create a class for your plugin inheriting from Plugin::
+Creating a plugin is simple. First of all you need to create necessary files: file for your plugin code (in our case: ``example_plugin.py``), ``setup.py`` and
```
Comment (DavidAndreev) [inline docs/source/index.rst]: I think there should be a coma : First of all, you need ..

**Comment by meluru** [comment]
Updated

File: docs/source/index.rst [hunk 25]
```diff
@@ -0,0 +1,302 @@
...
+
+.. autoclass:: Plugin
+    :members: init
+
+    .. automethod:: plugin_context()
+    .. classmethod:: instance()
+
+        The Plugin instance used by the current app
+
+    .. classmethod:: title()
```
Comment (ThiefMaster) [inline docs/source/index.rst]: Does it show up as `title()` or `title` in the docs? Because it's a class-level property and using `()` wouldn't work for obvious reasons ;)

File: example/app/app.py [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+plugin_engine = PluginEngine(app=app)
+plugin_engine.load_plugins(app)
+active_plugins = plugin_engine.get_active_plugins(app=app).values()
+
+for plugin in active_plugins:
+    with plugin.plugin_context():
+        app.register_blueprint(plugin.get_blueprint())
+
+
+@app.route("/")
```
Comment (ThiefMaster) [inline example/app/app.py]: single quotes

File: example/app/app.py [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+def hello():
+    return render_template('index.html', plugins=active_plugins)
+
+
+if __name__ == "__main__":
```
Comment (ThiefMaster) [inline example/app/app.py]: single quotes

File: flask_pluginengine/plugin.py [hunk 28]
```diff
@@ -119,12 +119,14 @@ def instance(cls):
     @classproperty
     @classmethod
     def title(cls):
+        """Plugin's title from the docstring"""
```
Comment (ThiefMaster) [inline flask_pluginengine/plugin.py]:
```
"""The title of the plugin.

Automatically retrieved from the docstring of the plugin class.
"""
```

File: flask_pluginengine/plugin.py [hunk 29]
```diff
@@ -119,12 +119,14 @@ def instance(cls):
...
         parts = trim_docstring(cls.__doc__).split('\n', 1)
         return parts[0].strip()
 
     @classproperty
     @classmethod
     def description(cls):
+        """Plugin's description from the docstring"""
```
Comment (ThiefMaster) [inline flask_pluginengine/plugin.py]:
```
"""The description of the plugin.

Automatically retrieved from the docstring of the plugin class.
"""
```

**Comment by ThiefMaster** [comment]
Would it make sense to split the docs into multiple pages? When you want to write a plugin for some application you probably don't care about how to use this extension in your own app. OTOH, an application should have its own documentation on how to write plugins for it.

File: docs/source/index.rst (hunk 25 above)
Comment (ThiefMaster) [inline docs/source/index.rst]: Just checked and it indeed shows up wrong.

**Comment by meluru** [comment]
Updated
